include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -377,45 +386,72 @@ __radix_sort_count_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments, :
 // radix sort: scan kernel (per iteration)
 //-----------------------------------------------------------------------
 
-template <typename _KernelName, ::std::uint32_t __radix_bits, typename _ExecutionPolicy, typename _CountBuf>
+template <typename _KernelName1, typename _KernelName2, ::std::uint32_t __radix_bits, typename _ExecutionPolicy,
+          typename _CountBuf
+#if _ONEDPL_COMPILE_KERNEL
+          ,
+          typename _Kernel1
+#endif
+          >
 sycl::event
-__radix_sort_scan_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments, _CountBuf& __count_buf,
-                         sycl::event __dependency_event)
+__radix_sort_scan_submit(_ExecutionPolicy&& __exec, ::std::size_t __scan_wg_size, ::std::size_t __segments,
+                         _CountBuf& __count_buf, sycl::event __dependency_event
+#if _ONEDPL_COMPILE_KERNEL
+                         ,
+                         _Kernel1& __kernel_1
+#endif
+)
 {
-    // typedefs
     using _CountT = typename _CountBuf::value_type;
 
-    // radix states used for an array storing bucket state counters
-    const ::std::uint32_t __radix_states = __get_states_in_bits(__radix_bits);
-
     auto __count_rng =
         oneapi::dpl::__ranges::all_view<_CountT, __par_backend_hetero::access_mode::read_write>(__count_buf);
 
+    // there are no local offsets for the first segment, but the rest segments shoud be scanned
+    // with respect to the count value in the first segment what requires n + 1 positions
+    const ::std::size_t __scan_size = __segments + 1;
+
+    __scan_wg_size = ::std::min(__scan_size, __scan_wg_size);
+    const ::std::size_t __iters_per_witem = __get_roundedup_div(__scan_size, __scan_wg_size);
+    const ::std::size_t __count_size = __count_buf.get_count();
+
+    const ::std::uint32_t __radix_states = __get_states_in_bits(__radix_bits);
+    const ::std::size_t __global_scan_begin = __count_size - __radix_states;
```

Comment:
This is the only place where we use `__count_size`. I'd suggest to remove this extra variable